### PR TITLE
feat: document reference

### PR DESCRIPTION
**Pull request by tevariou**
## Description

* Add DocumentReference to cohort builder criteria

### CODE REVIEW

File: src/components/Cohort/CreationCohort/DiagramView/components/GroupCard/components/CriteriaRightPanel/DocumentsForm/DocumentsForm.tsx [hunk 1]
```diff
@@ -83,16 +83,16 @@ const TestGeneratedForm: React.FC<TestGeneratedFormProps> = (props) => {
               validationRules: {
                 required: 'Merci de renseigné une recherche'
```
Comment (Jasopaum) [inline src/components/Cohort/CreationCohort/DiagramView/components/GroupCard/components/CriteriaRightPanel/DocumentsForm/DocumentsForm.tsx]: é -> er 🤕

File: src/services/cohortCreation/index.ts [hunk 2]
```diff
@@ -16,10 +23,25 @@ const getPatients = memoize(
     if (!query) return []
     const response = await apiFhir.get<FHIR_API_Response<IPatient>>(query)
     const patients = getApiResponseResources(response)
-    return patients ? patients.map((patient) => patient.id) : []
+    return patients?.map((patient) => patient.id) ?? []
   }
 )
 
+const getPatientsFromDocuments = memoize(
```
Comment (Jasopaum) [inline src/services/cohortCreation/index.ts:30]: Could we do a single query with an `_include` or a `_has` instead of that?

Comment (tevariou) [reply src/services/cohortCreation/index.ts:30]: Can I with the operator? I tried without success

Comment (Jasopaum) [reply src/services/cohortCreation/index.ts:30]: Oh, you're right, maybe we need to implement something 🤔
I'll try to see if we can do that

Comment (simonvadee) [reply src/services/cohortCreation/index.ts:30]: we can't, sorry

File: src/services/cohortCreation/index.ts [hunk 3]
```diff
@@ -16,10 +23,25 @@ const getPatients = memoize(
...
+  async (query: string, filter: string): Promise<(string | undefined)[]> => {
```
Comment (simonvadee) [inline src/services/cohortCreation/index.ts]: can you rename the args to: documentsQuery and patientsFilter ? it's not obvious

File: src/services/cohortCreation/index.ts [hunk 4]
```diff
@@ -16,10 +23,25 @@ const getPatients = memoize(
...
+    const response = await apiFhir.get<FHIR_API_Response<IDocumentReference>>(query)
+    const documents = getApiResponseResources(response)
+    const patients = documents
```
Comment (simonvadee) [inline src/services/cohortCreation/index.ts]: rename to patientIds ?

File: src/services/cohortCreation/index.ts [hunk 5]
```diff
@@ -16,10 +23,25 @@ const getPatients = memoize(
...
+      ?.map((document) => last(document.subject?.reference?.split('/')))
+      .filter((patient) => !!patient)
```
Comment (simonvadee) [inline src/services/cohortCreation/index.ts]: is the filter really needed ?

File: src/services/cohortCreation/index.ts [hunk 6]
```diff
@@ -16,10 +23,25 @@ const getPatients = memoize(
...
+    if (!patients?.length) return []
+
+    const allPatients = await getPatients(`/Patient?${filter}`)
+    return patients?.filter((patient) => allPatients.includes(patient)) ?? []
+  },
+  (a, b) => a + b
```
Comment (simonvadee) [inline src/services/cohortCreation/index.ts]: what's this ?

File: src/services/cohortCreation/index.ts [hunk 7]
```diff
@@ -64,49 +84,76 @@ const createCohortGroup = async (jsonQuery: string): Promise<IGroup> => {
...
+  const criteriaGroup = query.request[0].criteria
+
+  // We record every requests needed to fetch the patients/documents that meet the cohort criteria
+  // and then we fetch the patients/documents once
+  const { patientQueries, documentQueries } = criteriaGroup.reduce<{
+    patientQueries: string[]
+    documentQueries: string[]
+  }>(
+    (acc, group) => {
+      return group.criteria.reduce<{
```
Comment (simonvadee) [inline src/services/cohortCreation/index.ts:103]: nit: don't need to use return here

File: src/services/cohortInfos.ts [hunk 8]
```diff
@@ -151,11 +151,13 @@ const fetchCohort = async (cohortId: string | undefined): Promise<CohortData | u
     const cohortResult: CohortData = {}
     const response = getApiResponseResources(
       await api.get<FHIR_API_Response<IEncounter | IPatient | IGroup>>(
-        `/Patient?_has:Group:member:_id=${cohortId}&_revinclude=Group:member&_revinclude=Encounter:patient`
+        `/Patient?_has:Group:member:_id=${cohortId}&_revinclude=Group:member&_revinclude=Encounter:patient&_count=10000`
```
Comment (simonvadee) [inline src/services/cohortInfos.ts:154]: _count=1000 is enough

File: src/services/cohortCreation/index.ts (hunk 7 above)
Comment (tevariou) [reply src/services/cohortCreation/index.ts:103]: my ide complains for some reason if I don't

File: src/services/cohortCreation/index.ts (hunk 6 above)
Comment (tevariou) [reply src/services/cohortCreation/index.ts]: a callback to build the cache key https://lodash.com/docs/4.17.15#memoize

File: src/services/cohortCreation/index.ts (hunk 5 above)
Comment (tevariou) [reply src/services/cohortCreation/index.ts]: I guess if the reference is undefined

File: src/services/cohortCreation/index.ts (hunk 7 above)
Comment (simonvadee) [reply src/services/cohortCreation/index.ts:103]: ok

File: src/services/cohortCreation/index.ts (hunk 6 above)
Comment (simonvadee) [reply src/services/cohortCreation/index.ts]: ok

File: src/services/cohortCreation/index.ts (hunk 5 above)
Comment (simonvadee) [reply src/services/cohortCreation/index.ts]: which should never happen... nevermind

**Review by simonvadee — APPROVED** [review]

File: src/services/cohortInfos.ts (hunk 8 above)
Comment (simonvadee) [reply src/services/cohortInfos.ts:154]: no ?

Comment (tevariou) [reply src/services/cohortInfos.ts:154]: I may go for a little bit more than 1000, 1500 ? It's also fetching groups

Comment (simonvadee) [reply src/services/cohortInfos.ts:154]: _count only affects the resource you're searching for, not the included results

Comment (simonvadee) [reply src/services/cohortInfos.ts:154]: so there will be max 500 patients for the demo

**Pull request merged**